tsweb/server.go

```diff
@@ -0,0 +1,345 @@
+// Copyright (c) 2022 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package tsweb
+
+import (
+	"context"
+	"crypto/tls"
+	"expvar"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"golang.org/x/crypto/acme/autocert"
+	"tailscale.com/metrics"
+	"tailscale.com/types/logger"
+)
+
+const defaultTimeout = 10 * time.Second
+
+// ServerConfig is the initial configuration of a Server, for
+// consumption by NewServer.
+type ServerConfig struct {
+	// Name is a human-readable, alphanumeric name for the HTTP
+	// server.
+	// The name is for internal use only, for things like naming cache
+	// directories, identifying the source of autocert emails,
+	// human-readable debug pages, ...
+	Name string
+	// Addr specifies the TCP address for the server to listen on, in
+	// the form "host:port". If Addr specifies port 443, TLS serving
+	// is automatically configured and a second listener is set up on
+	// port 80 to handle HTTP to HTTPS redirection. All other ports
+	// result in HTTP-only serving.
+	Addr string
+	// Handler is the HTTP.Handler to invoke to handle
+	// requests. Cannot be nil.
+	Handler http.Handler
+	// Hostnames is the list of hostnames that the server accepts for
+	// serving TLS requests.
+	Hostnames []string
```

Comment:
maybe something like https://pkg.go.dev/golang.org/x/crypto/acme/autocert#HostPolicy instead?

then callers can use funcs if needed but otherwise use helpers like https://pkg.go.dev/golang.org/x/crypto/acme/autocert#HostWhitelist ?

(just using autocert as example)

Reply:
Changed to HostPolicy. The line I'm trying to walk is avoiding too much boilerplate in common cases, and I pictured the common case as "here's the 1 hostname I have". Other servers could go monkey with autocert configuration by hand before starting the server. But `[]string{` vs `autocert.HostWhitelist` isn't that much extra boilerplate.